axlearn/common/base_layer.py

```diff
@@ -267,84 +269,101 @@ def _methods_to_wrap_for_auto_child_context(self) -> dict[str, Callable]:
             if not hasattr(BaseLayer, method)
         }
 
-    def dtype(self):
-        if self.config.dtype is not None:
-            return self.config.dtype
-        if self.parent is not None:
-            return self.parent.dtype()
-        return jnp.float32
+    def _wrap_methods_with_auto_child_context(
+        self, methods: dict[str, Callable]
+    ) -> dict[str, Callable]:
+        cfg: Module.Config = self.config
 
-    def param_init(self) -> param_init.Initializer:
-        init = getattr(self, "_param_init", None)
-        if init is not None:
-            return init
-        return self.parent.param_init()
+        if cfg.remat_spec is not None:
+            for method_name, method_fn in dict(methods).items():
+                methods[method_name] = self._maybe_wrap_with_remat(method_fn)
 
-    # pylint: disable-next=no-self-use
-    def _create_layer_parameter_specs(self) -> dict[str, ParameterSpec]:
-        """Subclasses can override this method to add layer parameters."""
-        return {}
+        return super()._wrap_methods_with_auto_child_context(methods)
 
-    def _call_thunk(self, *args, method_fn, **kwargs) -> Callable[[], Any]:
-        cfg: Module.Config = self.config
-        self.vlog(
-            3, "BaseLayer._call_thunk %s.%s: remat=%s", self.path(), method_fn, cfg.remat_spec
-        )
+    def _maybe_wrap_with_remat(self, method_fn: Callable) -> Callable:
+        """Maybe wrap `method_fn` with jax remat.
```

Comment:
I may have missed it, but where is the comment about when wrapping happens?

Reply:
It's in the docstring on the next line:

```
This is called from `BaseLayer._wrap_methods_with_auto_child_context`.
```

Reply:
(In the past ruoming has asked me to add a copy when doing this kind of iteration.)